Fix Cursor is not updated when an element position changes

AlexsanderDamaceno opened the pull request:
#### 1368da9dddedc19b324d0ea1818c5f7632b9bd59
<pre>
Fix Cursor is not updated when an element position changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=53340">https://bugs.webkit.org/show_bug.cgi?id=53340</a>

Reviewed by Alan Baradlay.

When an element is re-laid out, the cursor does not change its style until the mouse is moved because this triggers EventHandler::selectCursor.
This issue is resolved by calling a cursor update function to update the cursor type during document layout.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout):
* LayoutTests/fast/events/mouse-cursor-change-02-expected.txt: Added.
* LayoutTests/fast/events/mouse-cursor-change-02.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/291948@main">https://commits.webkit.org/291948@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3172b0b91b422bc9053cbc9407631ec1a8f32c9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94403 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13995 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3772 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99422 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44934 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96453 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14293 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22423 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72038 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29368 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97405 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10629 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52370 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10322 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2925 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44247 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80555 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101466 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21458 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15657 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81040 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21709 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81253 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80414 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20061 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24965 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2341 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14683 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21435 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26605 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21124 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24584 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/41c6b9e7436c89c3f45d6da92e45e837cc20d9cc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

nt1m commented:
This needs a test

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/f9ad8c16e6e2d84bf07d463fc195f1b0553ef7db)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90852 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10389 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45799 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92905 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10784 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18706 "Build is in progress. Recent messages:Running configuration") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93853 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/10784 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/45799 "Build is in progress. Recent messages:Running configuration") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/10784 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/45799 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/10784 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/45799 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/18706 "Build is in progress. Recent messages:Running configuration") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/45799 "Build is in progress. Recent messages:Running configuration") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/55/builds/45799 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/680f59b7d0d4b96021496731828fe45440a44b98. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/112cadba6d9639f3c28ccfc3f48775a6af71f545. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/24fc12cb1d03237ba66a74e14bac46fb035fb537. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dd2f9c3f33207e6386f5b60e74d178e7938ce9f4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/603c716d483c280dffa5d6380f2e7a7c2f279a16. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

Ahmad-S792 commented on LayoutTests/fast/events/mouse-cursor-change-02.html:
```diff
@@ -0,0 +1,63 @@
+        }
+    </style>
+</head>
+<script src="../../resources/js-test-pre.js"></script>
```
New tests should just use `js-test.js` instead of `-pre.js` and `-post.js`.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d628e58bfb3c363848ee3de6f3dcd4f799b09da4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

Ahmad-S792 commented on LayoutTests/fast/events/mouse-cursor-change-02.html:
```diff
@@ -0,0 +1,61 @@
+        } 
+    </script>
+</body>
+</html>
```
No opening <html> available.

alanbaradlay replied:
instead of a 1s timer, I think we should just wait until the next rendering update. You can find many examples of how to do that in LayoutTests (UIHelper.renderingUpdate)

Ahmad-S792 commented:
> @nt1m the test that I added is failling only for **wpe-wk2** and **win-tests**, do have any ideia why?

For `wpe`, I would suggest asking in `gtk` channel on WebKit Open Source Slack. While for Windows, @fujii or @donny-dont might have any insight.

Ahmad-S792 commented:
@AlexsanderDamaceno - On `WPE`, other curosr tests are also marked as failure - https://github.com/WebKit/WebKit/blob/3009c67d8bd8fb9da9b48ed5df2f66432cdb6b74/LayoutTests/platform/wpe/TestExpectations#L532, so I wouldn't be surprise that it might be hitting same issue.

Ahmad-S792 commented:
> @AlexsanderDamaceno - On `WPE`, other curosr tests are also marked as failure -
> 
> https://github.com/WebKit/WebKit/blob/3009c67d8bd8fb9da9b48ed5df2f66432cdb6b74/LayoutTests/platform/wpe/TestExpectations#L532
> 
> , so I wouldn't be surprise that it might be hitting same issue and need to be marked similarly.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b30560eb77eefa7c06adbbd3f56756126034594c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

alanbaradlay commented on LayoutTests/fast/events/mouse-cursor-change-02.html:
```diff
@@ -0,0 +1,61 @@
+            var cursorInfo = window.internals.getCurrentCursorInfo(document);
+            setTimeout(() => {
+                verifyCursor();
+            }, 2000);       
```
:( This is bound to be flaky (2 sec timeout).

Ahmad-S792 replied:
requestAnimationFrame (rAf)?

alanbaradlay commented:
> > Do you know why RenderElement::styleDidChange does not cover this case?
> 
> I think styleDidChange is updating correctly the style information, but the function that select the type of cursor to display based on current style, only trigger when we move the mouse, If the element changes the position this do not trigger this function, even If element get the new style and is relayout


Ok, so it looks like `styleDidChange` only schedules cursor update when _cursor_ _style_ changes (which totally makes sense) and this PR is about when content changes under the cursor due to style mutation(s) unrelated to cursor.
While it is difficult to tell if layout has updated content _under_ the cursor, this PR introduces an unconditional rendering schedule, which may very well be okay, but I'd like to hear @smfr's opinion on this.

smfr commented:
I think this is a reasonable approach, but we need to measure if this additional cursor update per rendering update (which requires a new hit-test) is a performance or power regression.

cdumez commented:
> @cdumez could set up a perf this for test this changes impact?

Ok. I have started Speedometer A/B testing.

cdumez commented:
> > @cdumez could set up a perf this for test this changes impact?
> 
> Ok. I have started Speedometer A/B testing.

This is performance neutral on Speedometer 3.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5ae3094e0a1a4a330e8b3dde907bb9ea95512360. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1dbc490731340b538000739e0a1cde8b523fcf95. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a3172b0b91b422bc9053cbc9407631ec1a8f32c9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40887

AlexsanderDamaceno committed:
Fix Cursor is not updated when an element position changes
https://bugs.webkit.org/show_bug.cgi?id=53340

Reviewed by Alan Baradlay.

When an element is re-laid out, the cursor does not change its style until the mouse is moved because this triggers EventHandler::selectCursor.
This issue is resolved by calling a cursor update function to update the cursor type during document layout.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout):
* LayoutTests/fast/events/mouse-cursor-change-02-expected.txt: Added.
* LayoutTests/fast/events/mouse-cursor-change-02.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:

Canonical link: https://commits.webkit.org/291948@main

webkit-commit-queue commented:
Committed 291948@main (1368da9ddded): <https://commits.webkit.org/291948@main>

Reviewed commits have been landed. Closing PR #40887 and removing active labels.
